adds support for crypto.Signer in asymmetric signing methods

nmaki opened the pull request.

dgrijalva commented on ecdsa.go:
```diff
@@ -65,11 +71,18 @@ func (m *SigningMethodECDSA) Verify(signingString, signature string, key interfa
+		}
 	default:
-		return ErrInvalidKeyType
+		return ErrInvalidKey
```
This should still be key type, no?

nmaki replied:
yes that makes more sense. I'll update.

nmaki commented:
@dgrijalva I have updated the error type. 

rsa_pss was previously returning an ErrInvalidKey when the key was not an rsa key even before my updates. I've updated that to match the behavior of rsa and ecdsa. I can revert that back to its previous return behavior but assume consistency is better. Relevant line below

https://github.com/dgrijalva/jwt-go/pull/236/files#diff-5f1c612a96141d77f7ca41aa913cff64R94

hellais commented:
I would really like to see this feature land in the next version of jwt-go. It would make working with JWT and hardware tokens (like the yubikey 4) much easier thanks to https://github.com/thalesignite/crypto11.

What is the timeline for releasing 4.0.0? Is there something I can do to help out and make sure this feature makes it's way into the next release?

dgrijalva commented:
This is marked for the 4.0.0 release. It will make it in. I don’t remember if I’ve landed it on the branch yet, but I know I worked on it. I think this and the transition to option functions are the biggest changes for v4, so I want to make sure I understand everything in them.

I’m in the middle of moving, but I hope to get 4.0.0 out soon after we get settled. The major versions take a bit more time because I try to break the API as infrequently as possible, so I spend a good deal of time on them. 

Feedback on the release branch is always valuable.

-dave

> On Apr 11, 2018, at 3:06 AM, Arturo Filastò <notifications@github.com> wrote:
> 
> I would really like to see this feature land in the next version of jwt-go. It would make working with JWT and hardware tokens (like the yubikey 4) much easier thanks to https://github.com/thalesignite/crypto11.
> 
> What is the timeline for releasing 4.0.0? Is there something I can do to help out and make sure this feature makes it's way into the next release?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

hellais commented:
> This is marked for the 4.0.0 release. It will make it in.

This is great to hear. Thanks so much!

> I don’t remember if I’ve landed it on the branch yet, but I know I worked on it.

I don't see it inside of https://github.com/dgrijalva/jwt-go/pull/248. Is there some other branch I should be looking at?

> I think this and the transition to option functions are the biggest changes for v4, so I want to make sure I understand everything in them.

Yes this makes sense.

> Feedback on the release branch is always valuable.

I would be happy to help out in testing the 4.0.0 release branch with my codebase as soon as this branch is merged into it. I am currently testing it with a fork of my own, but I can switch to using the upstream 4.x.x branch.

Thank again for working on this great library!

azdagron commented:
`crypto.Signer` support is also a requirement for JWT signing in the https://github.com/spiffe/spire codebase. We are merging our own workaround for now but are very excited to leverage this PR when it gets merged.

vtolstov commented:
any news to get his merged to 4.x branch?

dgrijalva commented:
I can do so today. I've been distracted trying to figure out the best way to incorporate support for xerrors without requiring the package.

Just so I'm clear, the benefit here is being able to use some external system for signing? I see we're type checking against public keys. Is that likely to be brittle in any way we need to care about?

dgrijalva commented:
Just a reminder that the 4.x API isn't solidified yet. It will be once this is released. In the meantime, you probably want to use a package manager to pin down a specific version so it doesn't break between builds.

vtolstov commented:
i'm already have go.mod.so pin is not problem, for my use case i want to utilize x509 certificates for signing (ECDSA)
Global use case - sign token with ecdsa cert and use this token in service that tries to connect to some server with this token and check server cert fingerprint with fingerprint stored in token.
So i'm try to defense MITM for such case if somebody wants to try create fake server to steal token.

dgrijalva commented:
This has landed on `release_4_0_0`
